language/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -0,0 +1,521 @@
+            bail!(
+                "Name of dependency declared in package '{}' \
+                 does not match dependency's package name '{}'",
+                dep_name,
+                manifest.package.name,
+            )
+        }
+
+        let kind = dep.kind.clone();
+        self.extend_graph(kind, &manifest, skip_fetch_latest_git_deps, progress_output)
+            .with_context(|| format!("Resolving dependencies for package '{}'", dep_name))
+    }
+
+    /// Check that every dependency in the graph, excluding the root package, is present in the
+    /// package table.
+    fn check_consistency(&self) -> Result<()> {
+        for package in self.package_graph.nodes() {
+            if package == self.root_package {
+                continue;
+            }
+
+            if self.package_table.contains_key(&package) {
+                continue;
+            }
+
+            let dependees: Vec<_> = self
+                .package_graph
+                .neighbors_directed(package, Direction::Incoming)
+                .map(|pkg| String::from(pkg.as_str()))
+                .collect();
+
+            bail!(
+                "No source found for package {}, depended on by: {}",
+                package,
+                dependees.join(", "),
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Check that there isn't a cycle between packages in the dependency graph.  Returns `Ok(())`
+    /// if there is not, or an error describing the cycle if there is.
+    fn check_acyclic(&self) -> Result<()> {
+        let mut cyclic_components = algo::kosaraju_scc(&self.package_graph)
+            .into_iter()
+            .filter(|scc| scc.len() != 1 || self.package_graph.contains_edge(scc[0], scc[0]));
+
+        let Some(scc) = cyclic_components.next() else {
+            return Ok(())
+        };
+
+        // Duplicate start of the node at end for display
+        // SAFETY: Strongly connected components can't be empty
+        let mut cycle: Vec<_> = scc.iter().map(Symbol::as_str).collect();
+        cycle.push(cycle[0]);
+
+        bail!("Found cycle between packages: {}", cycle.join(" -> "));
+    }
+
+    /// Add the transitive closure of `DependencyMode::Always` edges reachable from the root package
+    /// to the `always_deps` set.  Assumes that if a package is already in the graph's `always_deps`
+    /// set, then the sub-graph reachable from it has already been explored.
+    fn discover_always_deps(&mut self) {
+        let mut frontier = vec![self.root_package];
+        while let Some(package) = frontier.pop() {
+            let new_frontier = self.always_deps.insert(package);
+            if !new_frontier {
+                continue;
+            }
+
+            frontier.extend(
+                self.package_graph
+                    .edges(package)
+                    .filter_map(|(_, dep, mode)| {
+                        (mode == &DependencyMode::Always).then_some(dep)
```

Comment:
I always find these sort of things easier to read with a `filter` then a `map`, make of that what you will (I don't really care one way or another)